Sources/SignalProducer.swift

```diff
@@ -1830,17 +1833,43 @@ extension SignalProducerProtocol {
 		let lifetime = Lifetime(lifetimeToken)
 
 		let state = Atomic(ReplayState<Value, Error>(upTo: capacity))
+		let start = Atomic(true)
 
-		let start: Atomic<(() -> Void)?> = Atomic {
-			// Start the underlying producer.
+		// Start the underlying producer.
+		func startHandler() {
 			self
+				.materialize()
 				.take(during: lifetime)
```

Comment:
The lifetime is what causes the failing test cases. The token tracks all outstanding producers, and breaks the observation only if there is none of them left.